### ISSUE TITLE
Non-htaccess support

### ISSUE THREAD

**Issue by sageworksstudio** [silverstripe/silverstripe-routewhitelist#16]
Feature Request

Since the whole point of routewhitelist is to reduce server load, would you consider making routewhitelist _not_ rely on .htaccess? Understandably, developers that develop on shared servers have to accept .htaccess as a fact of life, but for those of us that develop on controlled server .htaccess simply is the wrong way of doing things.

Thanks!

**Comment by candidasa** [comment]
Hi, can you explain why htaccess doesn't work on a controlled server? I chose to use htaccess because it gives very good performance when doing the kind of "does this file exist" lookup that module enables. The same lookup could be done using a simple PHP script, but you'd still need to configure the htaccess to route requests to that script instead of to SilverStripe. So why not just use the htaccess file?

**Comment by sageworksstudio** [comment]
Hi,

Turning on `AllowOverride` will cause Apache to evaluate .htaccess on every request. And not just every request, _every directory_, for every request. So the issue can scale exponentially.

From [Apache](https://httpd.apache.org/docs/current/howto/htaccess.html#when):

_There are two main reasons to avoid the use of .htaccess files._

_The first of these is performance. When AllowOverride is set to allow the use of .htaccess files, httpd will look in every directory for .htaccess files. Thus, permitting .htaccess files causes a performance hit, whether or not you actually even use them! Also, the .htaccess file is loaded every time a document is requested._

Thanks,
Chris

**Comment by candidasa** [comment]
Hi Chris,
Unfortunately, while Apache is, of course, correct to caution against using AllowOverwrite, SilverStripe itself relies on the htaccess file for access control of various folders, serving of assets from the assets directory and the redirect to the main.php file for invoking the CMS. Before making the routewhitelist module not rely on htaccess we would first need to rewrite SilverStripe not to rely on the htaccess file. So, while I appreciate the suggestion, I don't think it's realistic in this case.

**Comment by sageworksstudio** [comment]
SilverStripe actually doesn't rely on .htaccess. All of SilverStripe's directives fit perfectly in an Apache conf file.

```
<VirtualHost *:80>
    DocumentRoot /var/www/html/example
    ServerName example.com
    ServerAlias www.example.com

    # Route errors to static pages automatically generated by SilverStripe
    ErrorDocument 404 /assets/error-404.html
    ErrorDocument 500 /assets/error-500.html
    ### SILVERSTRIPE START ###
        <Directory /var/www/html/example/assets>
            <FilesMatch "\.(php|php3|php4|php5|phtml|inc)$">
                Deny from All
            </FilesMatch>
        </Directory>
        <Directory /var/www/html/example>
            # Deny access to templates (but allow from localhost)
            <Files *.ss>
                Order Deny,Allow
                Deny from All
                Allow from 127.0.0.1
            </Files>
            # Deny access to IIS configuration
            <Files web.config>
                Order Deny,Allow
                Deny from All
            </Files>
            # Deny access to YAML configuration files which might include sensitive information
            <Files *.yml>
                Order Allow,Deny
                Deny from All
            </Files>
            # Turn off index.php handling request issues with SilverStripe. (apache >=2.4)
            <IfModule mod_dir.c>
                DirectoryIndex disabled
            </IfModule>
            <IfModule mod_rewrite.c>
                SetEnv HTTP_MOD_REWRITE On
                RewriteEngine On

                #routewhitelist: send known invalid URLs straight to the 404 error page
                RewriteCond %{REQUEST_FILENAME} !-f
                RewriteCond %{DOCUMENT_ROOT}/routewhitelistcache/.htaccess -f
                RewriteCond %{REQUEST_URI} ^\/(.+?)(\/.*|\s*)$
                RewriteCond %{DOCUMENT_ROOT}/routewhitelistcache/%1 !-f 
                RewriteRule  .* assets/error-404.html [L,R=404]

                # Deny access to potentially sensitive files and folders
                RewriteRule ^vendor(/|$) - [F,L,NC]
                RewriteRule silverstripe-cache(/|$) - [F,L,NC]
                RewriteRule composer\.(json|lock) - [F,L,NC]
                # Process through SilverStripe if no file with the requested name exists.
                # Pass through the original path as a query parameter, and retain the existing parameters.
                RewriteCond %{REQUEST_URI} ^(.*)$
                RewriteCond %{REQUEST_FILENAME} !-f
                RewriteRule .* %{DOCUMENT_ROOT}/framework/main.php?url=%1 [QSA]
            </IfModule>
        </Directory>
    ### SILVERSTRIPE END ###
</VirtualHost>
```

**Comment by candidasa** [comment]
Ah, ok, I didn't realise that. Good to know.

It seems simple enough to add the whitelist config htaccess snippet to this Apache config. How do you think that is best represented in the codebase/documentation?

**Comment by sageworksstudio** [comment]
Well...admittedly I was only considering myself when I suggested this, but I guess it still has to be usable by developers that don't have root access to the Apache file.

That said, I think all you would have to do actually, is modify WhitelistGenerator's ensureWhitelistFolderExists() method to make the cache directory not world writable.:

`chmod($dir,0770);`

Then the module wouldn't rely on .htaccess at all, but of course developers that do rely on .htaccess can still use it too.

Developrs that do use Apache conf files could then just add the rewrite rules inside the `<IfModule mod_rewrite.c>` as long as they remember to enclose that inside the `<Directory /var/www/html/example>` See my modified example above.

**Comment by candidasa** [comment]
I don't understand how the write permissions of the cache directory impact the ability to use Apache virtual host configurations. Could you please explain?

**Comment by sageworksstudio** [comment]
Sorry, I think I'm conflating two issues. I guess ultimately you don't _need_ htaccess to use this module. You could always override the htaccess directives in a conf file. Because of that I would reword the requirements section of the readme to reflect that.

However, that said, you also don't need to protect the cache directory with htaccess if you just write it 0770 in the first place.

**Comment by candidasa** [comment]
Sorry for the intermittent updates. 

Regarding the permissions on the cache directory: what is your concern here? Is it that website visitors can potentially view the list of cached filenames?

**Comment by sageworksstudio** [comment]
Maybe I'm just paranoid, but creating a directory with 0777 permissions seems to be a security issue. _Reading_ files in the cache would be the least of my concerns. _Writing_ files to that directory and giving execute permissions, on the other hand, _could be_ a huge security hole.
Unless an application has a very specific reason to do so, giving the world write permission on a directory at the root of an application seems dangerous.
At the very least I would think creating the directory with 0774 permissions would be safest and should still work in a shared hosting environment.

Regards,
Chris
